### PR TITLE
Think-MD integration improvements

### PR DESCRIPTION
**Pull request by zzainulabidin**

### CODE REVIEW

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java [hunk 1]
```diff
@@ -132,29 +131,27 @@ public static boolean isChild(String baseEntityID) {
         return res.get(0) > 0;
     }
 
-    public static String getBaseEntityID(String identifierType, String id) {
-        String selection = "select base_entity_id from ec_child where ? = ?";
-        String[] selectionArgs = new String[]{identifierType, id};
```
Comment (githengi) [inline opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java:137]: Why was this removed, is this not being used by others.

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java [hunk 2]
```diff
@@ -132,29 +131,27 @@ public static boolean isChild(String baseEntityID) {
...
+    public static String getBaseEntityID(String thinkMdId) {
+        String selection = " thinkmd_id = ? ";
+        String[] selectionArgs = new String[]{thinkMdId};
         return ChildDao.queryColumnWithEntityId(selection, selectionArgs, BASE_ENTITY_ID);
     }
```
Comment (githengi) [inline opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java:138]: shouldn't this reuse the older method `getBaseEntityID(String identifierType, String id)` and pass the params `getBaseEntityID("thinkmd_id", thinkMdId)`

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java [hunk 3]
```diff
@@ -132,29 +131,27 @@ public static boolean isChild(String baseEntityID) {
...
 
+    public static String getThinkMDId(String childBaseEntityId) {
+        String selection = " base_entity_id = ? ";
+        String[] selectionArgs = new String[]{childBaseEntityId};
+        return ChildDao.queryColumnWithEntityId(selection, selectionArgs, THINK_MD_ID);
+    }
```
Comment (githengi) [inline opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java:144]: shouldn't this reuse the older method getBaseEntityID(String identifierType, String id) and pass the params getBaseEntityID("base_entity_id", childBaseEntityId)

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java [hunk 4]
```diff
@@ -482,16 +486,24 @@ public void createCarePlanEvent(@NotNull Context context, @NotNull String encode
...
-                if (baseEntityId != null) {
+                if (childBaseEntityId != null) {
                     Event carePlanEvent = ThinkMDLibrary.getInstance().createCarePlanEvent(encodedBundle,
                             getFormTag(getAllSharedPreferences()),
-                            baseEntityId);
+                            childBaseEntityId);
+                    updateLocalStorage(childBaseEntityId, "thinkmd_fhir_bundle", encodedBundle);
+
+                    for (Obs obs : carePlanEvent.getObs()) {
+                        if (StringUtils.isEmpty(obs.getFormSubmissionField())) continue;
```
Comment (githengi) [inline opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java:498]: do we really need this statement since the one below checks for a specific value

**Review by githengi — CHANGES_REQUESTED** [review]

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java [hunk 5]
```diff
@@ -502,6 +514,18 @@ public void createCarePlanEvent(@NotNull Context context, @NotNull String encode
         appExecutors.diskIO().execute(runnable);
     }
 
+    private void updateLocalStorage(String childBaseEntityId, String fieldName, String fieldValue) {
+        if (getChildAllCommonsRepository() != null) {
+            ContentValues values = new ContentValues();
+            values.put(fieldName, fieldValue);
+            getChildAllCommonsRepository().update("ec_child", values, childBaseEntityId);
+        }
```
Comment (githengi) [inline opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java:522]: instead of using this custom logic whey don't we use ec_client_fields to extract this value. That is better since it also works for events synced from server

**Review by githengi — CHANGES_REQUESTED** [review]

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java (hunk 1 above)
Comment (zzainulabidin) [reply opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java:137]: This method is added again.

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java (hunk 2 above)
Comment (zzainulabidin) [reply opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java]: Done.

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java (hunk 1 above)
Comment (zzainulabidin) [reply opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java:137]: This method is added again.

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java (hunk 3 above)
Comment (zzainulabidin) [reply opensrp-chw-core/src/main/java/org/smartregister/chw/core/dao/ChildDao.java]: Done.

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java (hunk 4 above)
Comment (zzainulabidin) [reply opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java:499]: Yes we need this statement, we need to query the obs and update the local DB with the care plan date and care plan.

File: opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java (hunk 5 above)
Comment (zzainulabidin) [reply opensrp-chw-core/src/main/java/org/smartregister/chw/core/interactor/CoreChildProfileInteractor.java:523]: I am using both, ec_client_fields took some time to process the event and update the local DB, we immediately need to update the local DB so that, accurate date of the care plan and care plan showed when the user has Thinkmd assessment.

**Review by rkodev — APPROVED** [review]

**Review by githengi — APPROVED** [review]
